modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerConsistencyCountersTest.java

```diff
@@ -0,0 +1,575 @@
+     */
+    @Test
+    public void testCountersOnCrachRecovery() throws Exception {
+        int nodes = 3;
+        int backupNodes = nodes - 1;
+
+        IgniteEx ignite = startGrids(nodes);
+
+        ignite.cluster().state(ClusterState.ACTIVE);
+
+        IgniteCache<Object, Object> cache = ignite.createCache(new CacheConfiguration<>()
+            .setAffinity(new RendezvousAffinityFunction(false, 1))
+            .setBackups(backupNodes)
+            .setName(DEFAULT_CACHE_NAME)
+            .setAtomicityMode(atomicityMode)
+            .setWriteSynchronizationMode(FULL_SYNC) // Allows to be sure that all messages are sent when put succeed.
+            .setReadFromBackup(true)); // Allows to check values on backups.
+
+        int updateCnt = 0;
+
+        // Initial preloading.
+        for (int i = 0; i < 2_000; i++) { // Enough to have historical rebalance when needed.
+            cache.put(i, i);
+
+            updateCnt++;
+        }
+
+        // Trick to have historical rebalance on cluster recovery (decreases percent of updates in comparison to cache size).
+        if (historical) {
+            stopAllGrids();
+            startGrids(nodes);
+        }
+
+        int preloadCnt = updateCnt;
+
+        Ignite prim = primaryNode(0L, DEFAULT_CACHE_NAME);
+        List<Ignite> backups = backupNodes(0L, DEFAULT_CACHE_NAME);
+
+        AtomicBoolean prepareBlock = new AtomicBoolean();
+        AtomicBoolean finishBlock = new AtomicBoolean();
+
+        AtomicReference<CountDownLatch> blockLatch = new AtomicReference<>();
+
+        TestRecordingCommunicationSpi.spi(prim).blockMessages(new IgniteBiPredicate<ClusterNode, Message>() {
+            @Override public boolean apply(ClusterNode node, Message msg) {
+                if ((msg instanceof GridDhtTxPrepareRequest && prepareBlock.get()) ||
+                    ((msg instanceof GridDhtTxFinishRequest || msg instanceof GridDhtAtomicSingleUpdateRequest) && finishBlock.get())) {
+                    CountDownLatch latch = blockLatch.get();
+
+                    assertTrue(latch.getCount() > 0);
+
+                    latch.countDown();
+
+                    return true; // Generating counter misses.
+                }
+                else
+                    return false;
+            }
+        });
+
+        int blockedFrom = reuseKeys ? 0 : preloadCnt;
+        int committedFrom = blockedFrom + 1_000;
+
+        int blockedKey = blockedFrom - 1; // None
+        int committedKey = committedFrom - 1; // None
+
+        List<String> backupMissed = new ArrayList<>();
+        List<String> primaryMissed = new ArrayList<>();
+
+        IgniteCache<Integer, Integer> primCache = prim.cache(DEFAULT_CACHE_NAME);
+
+        Consumer<Integer> cachePut = (key) -> primCache.put(key, -key);
+
+        Set<Thread> ths = ConcurrentHashMap.newKeySet();
+
+        Consumer<Integer> cachePutAsync = (key) -> {
+            Thread th = new Thread(() -> cachePut.accept(key));
```

Comment:
I'd suggest using `GridTestUtils#runAsync` since the thread will obey the test framework behaviour.

Reply:
Done